### DISCUSSION TITLE
Next Route API requests inside authorize function from Next Auth gets Unauthorized 401

### DISCUSSION THREAD

**Discussion by vitormarkis** [nextauthjs/next-auth#7093] [Help]
I've been struggling to create credentials based authentication using Next Auth, my auth data flow so far is:
1. client gather form data, and validates
2. the same function uses singIn function from nextauth using "credentials", and passing the fields
3. authorize function receives this data and makes a post request on "/api/login" next api route endpoint
4. authorize function returns the api response, which has the user session info and jwt token

The error occur when i invoke the submitHandler function (react-hook-form)

## Client
```tsx
const submitHandler: SubmitHandler<IUserLogin> = async (formData) => {
  try {
    const { email, password } = userLoginSchema.parse(formData)
    await signIn("credentials", { email, password, csrfToken, callbackUrl: "/testing", redirect: false })
  } catch (error) {
    console.log(error)
  }
}
```

## Authorize function
```tsx
const { email, password } = userLoginSchema.parse(credentials)

const reqValidation = z.object({
  user: userSchema,
  token: z.string(),
})

const { token, user } = await api
  .post("/auth/login", { email, password })
  .then((res) => reqValidation.parse(res.data))

return {
  ...user,
  repos_amount: null,
}
```

The error: `http://localhost:3000/api/auth/callback/credentials? 401 (Unauthorized)`

I noticed there is some csfrToken that i should be sending, but there is no explanation saying when should i send it or where.

**Comment by Manitej66** [comment]
You don't need to pass CSRF Token, nextauth automatically adds it. (read [docs](https://next-auth.js.org/getting-started/client#signin))

```
Using the signIn() method ensures the user ends back on the page they started on after completing
 a sign in flow. It will also handle CSRF Tokens for you automatically when signing in with email.
```

Just use like below
```js
await signIn("credentials", {
  callbackUrl: "/",
  redirect: false,
  email: data.email.trim().toLowerCase(),
  password: data.password,
});
```
